core/main/configuration.rb

```diff
@@ -76,6 +76,96 @@ def validate
         true
       end
 
+      #
+      # Returns the configuration value for the http server host
+      # If nothing is set it should default to 0.0.0.0 (all interfaces)
+      def local_host
+        get('beef.http.host') || '0.0.0.0'
+      end
+
+      #
+      # Returns the configuration value for the http server port
+      # If nothing is set it should default to 3000
+      def local_port
+        get('beef.http.port') || '3000'
+      end
+
+      #
+      # Return the local protocol
+      # if nothing is set default to http
+      def local_proto
+        get('beef.https.enabled') ? 'https' : 'http'
+      end
+
+      #
+      # Returns the configuration value for the local https enabled
+      # If nothing is set it should default to false
+      def local_https_enabled
+        get('beef.https.enabled') || false
+      end
+
+      #
+      # Returns the configuration value for the http server host
+      def public_host
+        get('beef.http.public')
+      end
+
+      #
+      # Returns the beef host which is used by external resources
+      # e.g. hooked browsers
+      def beef_host
+        public_host || local_host
+      end
+
+      #
+      # Returns the beef port which is used by external resource
+      # e.g. hooked browsers
+      def beef_port
+        public_port || local_port
+      end
+
+      def public_enabled?
+        !get('beef.http.public').nil?
+      end
+      
+      #
+      # Returns the beef protocol that is used by external resources
+      # e.g. hooked browsers
+      def beef_proto
+        if public_enabled? && public_https_enabled? then
+          return 'https'
+        elsif public_enabled? && !public_https_enabled?
+          return 'http'
+        elsif !public_enabled?
+          return local_proto
+        end
+      end
+
+      #
+      # Returns the beef scheme://host:port for external resources
+      # e.g. hooked browsers
+      def beef_url_str
+        "#{beef_proto}://#{beef_host}:#{beef_port}"
+      end
+
+      # Returns the configuration value for the http server port
+      # If nothing is set it should default to 3000
+      def public_port
+        return get('beef.http.public_port') unless get('beef.http.public_port').nil?
+
+        return '443' if public_https_enabled?
+        return '80' unless public_host.nil?
+
```

Comment:
This is set out in an interesting way, going from this:
```
      def beef_proto
        if public_enabled? && public_https_enabled? then
          return 'https'
        elsif public_enabled? && !public_https_enabled?
          return 'http'
        elsif !public_enabled?
          return local_proto
        end
```
to what you have in the public_port is a bit of a change.
Why it's set out like this?

Reply:
sup, thanks for looking of the draft.
so with public port there is an issue with people enabling public and not being able to set up https (ngrok)

when the person has not set `public_port` but has set `public` it will resolve to `local_port` with current code that is written throughout the application which is not always the case

https -> http (ngrok)

so for public port my assumption were that the following is the priority for public port

if there is a public_port set return that public port
if the user has enabled the new configuration variable `public_https_enabled` and not set a port then default it to 443
if public host is set and there is not https enabled then it should default to 80
the last condition would be if there are no public configuration variables set

beef_proto

first off it should be refactored to read better using `return XXX if ABC`

so the priority of beef_proto (the one that used in the hooked browsers) is as

1. if the public host has been enabled and the user has enabled the new variable `public_https_enabled` it should be `https`
2. if the user has set a public host and not enabled https then it should be `http`
3. if there are no public variables set then use the `local_proto`

I could have assumed the wrong priorities so let me know that can easily be fixed
Also refactoring as I go, to the `public_port` method can now use the new `public_enabled` method

keep up the reviewing much appreciated